### ISSUE TITLE
Limit traffic for a certain repo?

### ISSUE THREAD

**Issue by choldgraf** [jupyterhub/binderhub#242]
If you look at the distribution of where traffic comes from, it follows some kind of exponential distribution:

![image](https://user-images.githubusercontent.com/1839645/32452155-67c68062-c2de-11e7-860e-929f336c6624.png)

I'm a bit worried that at some point we'll hit a case where traffic is being totally dominated by a single popular repository, at the expense of all others. How do we think we should handle this? One thought I had was to limit the avg number of connections from one repo or something like. But we should discuss!

**Comment by ctb** [comment]
I think limiting it to something like ~50 max of any one repo would be fine.
Although of course one could just fork the repo into a new org or repo name
to evade...

**Comment by yuvipanda** [comment]
Operationally, this isn't a problem right now - we've tons of headroom! We can look at figuring this out when we do hit our limits. And even then, IMO we should try to find other solutions (more money?!) than try to do this kinda limiting.

**Comment by ctb** [comment]
I love your optimism and hopefulness @yuvipanda!

Hmm, could plan for a LIGO-style binder event... one will inevitably happen...

**Comment by choldgraf** [comment]
As an example of where this could get hairy, I'm going to merge a PR into the `sphinx-gallery` package that will auto-generate a binder link for any examples that it builds.

Matplotlib uses sphinx-gallery and I was planning to deploy this plugin on their documentation.

The Matplotlib website gets about 800,000 pageviews a week.

:-)

**Comment by yuvipanda** [comment]
Does it still not need users to click a button before they launch a binder?

**Comment by choldgraf** [comment]
oh for sure it'll be just a fraction of a percent of that amount...I guess we can cross that bridge when we come to it, but just giving a heads up that spikes in traffic may happen sooner than later

**Comment by yuvipanda** [comment]
+1.

We could also try limiting the following:

1. Idle timeout (currently ten minutes)
2. RAM allocated (currently 1G)

before starting to do hard limits.

**Comment by choldgraf** [comment]
yeah I'm +1 on idle timeout culling...though I think these changes can be reactionary once problems pop up

**Comment by willingc** [comment]
Just curious. Wouldn't this issue be better placed in the `mybinder-deploy` repo? It's based on the usage metrics of a particular deployment. If there is no next step for BinderHub development here, let's move it to the deployment repo.

**Comment by yuvipanda** [comment]
totally agree re: mybinder.org-deploy repo!

**Comment by choldgraf** [comment]
I put it here because I imagined it being some kind of configuration in the helm chart that could be used by other binderhub deployments too... But if people think it will be unique to our use case for the foreseeable future, I can move it. Lemme know!

**Comment by willingc** [comment]
I think for now this is specific to our use case @choldgraf.

There's a bunch of grey area in terms of what should go where. Here's how I am viewing it at the moment (since our plates runneth over):

- Things that are related to mybinder.org deployment get filed there first (capacity, sre, config for mybinder, etc.)
- If after triaging the issue in the mybinder-deploy repo, there is clear demand for a feature enhancement to binderhub or repo2docker or helm chart or jupyterhub, we move or open an issue there.

We could certainly create a bunch of nice to have features but it would be helpful to have the 'really should have' percolate up with demand from other users too.

**Comment by choldgraf** [comment]
> We could certainly create a bunch of nice to have features but it would be helpful to have the 'really should have' percolate up with demand from other users too.

this is a good point. As we have some kind of idea not to deal with this until it becomes an issue, shall I port this over to `mybinder.org`, or just close this one now and re-open an issue on `mybinder.org` if/when it becomes a problem?

**Comment by willingc** [comment]
Let's just close it here @choldgraf. Perhaps open an issue on `mybinder-deploy` with idle timeout culling if/when needed.

**Comment by willingc** [comment]
I'm going to close this and if it becomes a concern later with mybinder.org we can file and issue on `mybinder-deploy`.
